### DISCUSSION TITLE
Help with testing mimetypes on File Uploads

### DISCUSSION THREAD

**Discussion by Jubeki** [filamentphp/filament#14102] [Help]
### Package

Form builder

### Package Version

v3.2.109

### How can we help you?

Github Repository with all files and a failing test, which should pass:
https://github.com/Jubeki/filament-mimetype-testing/commit/2a3e360fbd591bab78a1613cdb58f988a2fff544

I try to test the behaviour of the `mimetype` validation rule for a failing mime type, but I always get that the form has no errors, even though it should have one.

Example: I allow `application/pdf` and `image/jpg`, but I supply a file with the mime type `text/plain` with a file ending `.jpg`.

How would you test this behaviour?

The test I wrote:

```php
test('example', function () {

    livewire(FileCollectionForm::class)
        ->set('data.files', [
            UploadedFile::fake()->create('file1.jpg', 1024, 'text/plain'),
        ])
        ->call('create')
        ->assertHasFormErrors([
            'data.files' => 'mimetypes'
        ]);
});
```

The file upload:

```php
class FileCollectionForm extends Component implements HasForms
{
    use InteractsWithForms;

    public ?array $data = [];

    public function mount(): void
    {
        $this->form->fill();
    }

    public function form(Form $form): Form
    {
        return $form
            ->schema([

                FileUpload::make('files')
                    ->multiple()
                    ->disk('local')
                    ->storeFileNamesIn('file_names')
                    ->required()
                    ->acceptedFileTypes(['application/pdf', 'image/jpg', 'image/jpeg', 'image/png'])
                    ->maxSize(5120)
                    ->maxFiles(5)

            ])
            ->statePath('data');
    }

    public function create(): void
    {
        FileCollection::create($this->form->getState());
    }

    public function render(): View
    {
        return view('livewire.file-collection-form');
    }
}
```

**Comment by leandrocfe** [comment]
Maybe you need to set the file ext: `UploadedFile::fake()->create('file1.txt', 1024, 'text/plain')`

Not sure if `mimes` validation works, but you could use `->assertHasFormErrors(['files'])` to verify the entire file validation:

```php
livewire(FileCollectionForm::class)
    ->fillForm([
        'files' => UploadedFile::fake()->create('file1.txt', 1024, 'text/plain'),
    ])
    ->call('create')
    ->assertHasFormErrors(['files'])
```

**Reply by Jubeki** [reply]
While changing the file extensions makes the test pass, the intention of the test is to have a allowed file extension, but a not allowed file mime.

I tested also with your code (except file extension), but I still get a failing test.

**Comment by Jubeki** [comment]
Okay, it seems to be a problem with Livewire TemporaryUploadedFile, which will return the incorrect mime type based on the extension, instead of using the provided mime type. Not sure how I want to prevent that, but I will take a further look.

https://github.com/livewire/livewire/blob/597a2808d8d3001cc3ed5ce89a6ebab00f83b80f/src/Features/SupportFileUploads/TemporaryUploadedFile.php#L55-L68

```php
    public function getMimeType(): string
    {
        $mimeType = $this->storage->mimeType($this->path);


        // Flysystem V2.0+ removed guess mimeType from extension support, so it has been re-added back
        // in here to ensure the correct mimeType is returned when using faked files in tests
        if (in_array($mimeType, ['application/octet-stream', 'inode/x-empty', 'application/x-empty'])) {
            $detector = new FinfoMimeTypeDetector();


            $mimeType = $detector->detectMimeTypeFromPath($this->path) ?: 'text/plain';
        }


        return $mimeType;
    }
```
